core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -765,20 +777,26 @@ class GroupCoordinator(val brokerId: Int,
   }
 
   def tryCompleteHeartbeat(group: GroupMetadata, member: MemberMetadata, heartbeatDeadline: Long, forceComplete: () => Boolean) = {
-    group synchronized {
-      if (shouldKeepMemberAlive(member, heartbeatDeadline) || member.isLeaving)
-        forceComplete()
-      else false
+    val complete = group synchronized {
+      shouldKeepMemberAlive(member, heartbeatDeadline) || member.isLeaving
     }
+    if (complete)
+      forceComplete()
+    else
+      false
   }
 
   def onExpireHeartbeat(group: GroupMetadata, member: MemberMetadata, heartbeatDeadline: Long) {
+    var preparingRebalance = false
```

Comment:
I wonder if we should introduce a helper method that runs another function outside the lock based on the result of the original function. Not sure if useful, just an idea.